### ISSUE TITLE
Closest

### ISSUE THREAD

**Issue by deber1980** [brentp/vcfanno#11]
Hello,

I wanted to ask if there is a way with vcfanno to annotate a variant with its proximity to a certain feature. Specifically, I would like to know if a variant is close to the intron/exon boundary. So the input would be a bed file with all the ranges of the exons and the output of the annotation would be the distance from the boundry for each variant.

Thanks

**Comment by brentp** [comment]
this is not currently available in `vcfanno`. The underlying intersection library supports closest, KNN, and other operations that aren't strictly intersection, but it's not exposed to `vcfanno`.

I've had a few people ask about this and I think it's a reasonable request. It will take some design to figure it out ...

making `closest` work with the parallel chrom-sweep will be more difficult because of the tabix queries. When a query chunk is acquired, it's bounds are used to create a tabix iterator of the annotation file. If the closest annotation interval is just before the lower bound (or just after the upper bound), it would be missed. 
If the tabix implementation had a way to ask for the bounds and the 1 (or K for k-neareast) up/down-stream interval this would be trivial to resolve.

then the user would specify a command-line something like `vcfanno -nearest $n` where `$n` indicates the number of nearest-neighbors to find. And we could have a new builtin operation called "distance" that would return a float of the distance.

**Comment by crazyhottommy** [comment]
Please add this feature, thanks very much!

**Comment by deber1980** [comment]
Many thanks Brent!

**Comment by brentp** [comment]
The simplest way to implement this would be to require a -maxdist parameter as well as the $n. e.g.

`vcfanno -nearest 4 -maxdist 100000`

and it would only look up to a maximum of 100000 bases away for the 4 nearest neighbors. @deber1980, @crazyhottommy would this meet your needs? 

The other option is to do vanilla (not parallel) chromsweep when -nearest is specified. This would ensure that all $n nearest neighbors are recovered, but wouldn't allow for parallelization.

**Comment by deber1980** [comment]
From my side the first solution you suggested (require a -maxdist parameter as well as the $n) meets my needs.
Thanks again

**Comment by crazyhottommy** [comment]
YES, `vcfanno -nearest 4 -maxdist 100000` works for me.
Thanks!

**Comment by crazyhottommy** [comment]
if it is not of too much work, could you please add options to restrict nearest genes upstream or downstream of the breakpoint? it has to do with the strandness of breakpoints https://github.com/crazyhottommy/DNA-seq-analysis/blob/master/vcf_and_bedpe_sv_format.md

I am currently using ChIPseeker to annotate the structural variants. see here https://github.com/GuangchuangYu/ChIPseeker/issues/17
and here https://github.com/GuangchuangYu/ChIPseeker/issues/23

My post http://crazyhottommy.blogspot.hk/2016/01/find-nearest-upstream-genes-using.html

In a word, I want to annotate the breakpoint with overlapping genes first (no matter the strandness of the gene and the breakpoint). If there is no overlapping gene for that breakpoint, then find the closest gene upstream of the breakpoint.

Thanks very much!
Ming

**Comment by brentp** [comment]
I think I have a way to do this. 
@crazyhottommy if you have 2 line example VCF and a short gene annotation file, I can debug further.

**Comment by crazyhottommy** [comment]
Please see attachment for the vcf file and below for the bed file. Thanks!

```
chr10 34048641 34061608        -     NR_038932
chr15 38226808 38243623        +    NM_152453
chr15 37091301 37110707       -     NR_027320
chr15 37183222 37393500       -     NM_001220482
chr10 34398488 35104253       -  NM_001184785
chr15 36871804 37102461       +  NM_001130010

```

breakpoint A should be annotated  by NM_001184785 ( in the intron of this gene)
breakpointB should be annotated by  NM_001220482.  (upstream of breakpointB)
[example.vcf.gz](https://github.com/brentp/vcfanno/files/88099/example.vcf.gz)

Thanks!
Ming

**Comment by brentp** [comment]
so you're using "upstream" based on the strand of the BND event, right? not on the gene?

**Comment by brentp** [comment]
I have this working modulo understanding your strand issues. If you let me know about that, I'll post a full example and add it to the docs as it's quite useful.

**Comment by crazyhottommy** [comment]
@brentp yes, upstream based on the strand of the BND event. Thanks very much!

**Comment by crazyhottommy** [comment]
Note that some breakpoints (at the start or end of a chromosome) will not have a closest upstream gene.

**Comment by brentp** [comment]
so if the ID ends in _1 (e.g. 6998_1), pull the first strand from STRANDS
and if it ends in _2, pull the 2nd? then use that to determine up/downstream?

**Comment by crazyhottommy** [comment]
That's what I think. However, please look at Colby Chiang's comments on my question  as well https://github.com/hall-lab/svtools/issues/5

My vcf is from lumpy output. I do think the strandness makes sense if you look at my note here https://github.com/crazyhottommy/DNA-seq-analysis/blob/master/vcf_and_bedpe_sv_format.md#please-also-check-the-figure-from-structural-variation-discovery-in-the-cancer-genome-using-next-generation-sequencing-computational-solutions-and-perspectives

`+` and `-` represents the strandness of the supporting reads. see here as well https://github.com/svviz/svviz/issues/28

**Comment by crazyhottommy** [comment]
if you look at this particular example:

```
8       9174263 5956_1  N       N[14:38454918[  25.35   .       SVTYPE=BND;STRANDS=+-:4;CIPOS=-9,8;CIEND=-9,8;CIPOS95=0,0;CIEND95=0,0;MATEID=....

14      38454918        5956_2  N       ]8:9174263]N    25.35   .       SVTYPE=BND;STRANDS=-+:4;CIPOS=-9,8;CIEND=-9,8;CIPOS95=0,0;CIEND95=0,0

```
when converted to bedpe:

```
8   9174253 9174271 14  38454908    38454926    my_sv   25.35  +   -  BND  ....
```

`_1` is `+` `_2` is `-`

So it looks like need to pull the first strand from SRANDS no matter the ID ends in _1 or _2, because the STRANDS changes from `+-` to `-+` in the record.

Ming
